### PR TITLE
Make covidhub_evaluation.Rmd more efficient when previous data exists

### PR DESCRIPTION
**Pull request by JedGrabman**
Description:
Checks for previously saved data before requesting / processing new data, leading to substantial speed improvements for weekly updates.

Change log:
- Checks for previous `predictions_cards` and `score_cards` before requesting or processing data.
- Only save `predictions_cards` and `score_cards`, not all values.
- Updated code for determining which dates to request (includes only desired dates, doesn't use NA).

### CODE REVIEW

File: R-notebooks/covidhub_evaluation.Rmd
Comment (krivard) [inline R-notebooks/covidhub_evaluation.Rmd:194]: Should this be computed dynamically?

File: R-notebooks/covidhub_evaluation.Rmd
Comment (sgsmob) [inline R-notebooks/covidhub_evaluation.Rmd:103]: Took me a while to parse through this.  This variable stores all dates except for the last one in each continuous blocks?  I would like to see a more descriptive name, like `previous_{consecutive_}dates_with_forecasts`.

File: R-notebooks/covidhub_evaluation.Rmd
Comment (sgsmob) [inline R-notebooks/covidhub_evaluation.Rmd:97]: Would find this easier to read as `comparable_forecast_dates`

File: R-notebooks/covidhub_evaluation.Rmd
Comment (sgsmob) [inline R-notebooks/covidhub_evaluation.Rmd:92]: Can you leave a comment explaining exactly what is "new" about the `new_dates`?  They are dates we don't have a comparable prediction for?

File: R-notebooks/covidhub_evaluation.Rmd
Comment (sgsmob) [inline R-notebooks/covidhub_evaluation.Rmd:107]: Is this equivalent to `dates_comp = as_date(setdiff(dates_comp, seen_dates[[forecaster_idx]]))`?  If so, I think it would be more readable and probably more efficient the way I wrote it.

File: R-notebooks/covidhub_evaluation.Rmd
Comment (sgsmob) [inline R-notebooks/covidhub_evaluation.Rmd:134]: Do you want `i %in% old_forecaster_idxs` instead?  My understanding is that `old_forecaster_idxs` is a list of index positions such as `c(1, 3, 7, 9)` corresponding to forecasters already seen.  So `old_forecaster_idxs[i]` asks for the `i`th member of the list rather than whether `i` is in the list.

File: R-notebooks/covidhub_evaluation.Rmd
Comment (sgsmob) [inline R-notebooks/covidhub_evaluation.Rmd:139]: Same comment about `old_forecaster_idxs[i]`.

**Review by sgsmob — CHANGES_REQUESTED** [review]

File: R-notebooks/covidhub_evaluation.Rmd
Comment (JedGrabman) [reply R-notebooks/covidhub_evaluation.Rmd:103]: >This variable stores all dates except for the last one in each continuous blocks

Yes, but the reason is a little bit more subtle.

CMU makes forecasts once a week. We consider a forecast date to be comparable to a CMU forecast if it was made on the same day or the previous day. So, if an organization makes weekly forecasts on a Sunday or Monday, we want to include it.

But, some organizations make forecasts daily and we don't want to include both the Sunday and Monday versions in that case, just the Monday version.

So, it is the last day in the chain, but it is really just excluding the Sunday version if the Monday version is also there.

File: R-notebooks/covidhub_evaluation.Rmd
Comment (JedGrabman) [reply R-notebooks/covidhub_evaluation.Rmd:107]: Agreed. That looks equivalent and clearer.

File: R-notebooks/covidhub_evaluation.Rmd
Comment (JedGrabman) [reply R-notebooks/covidhub_evaluation.Rmd:134]: The result from `match` is the same length as `forecasters` and specifies which value in the 2nd vector matches the value in the 1st vector (with NA representing no match).  

For example, `match(c('foo', 'bar', 'baz'), c('baz', 'foo'))` would be `c(2, NA, 1)`.

so, `old_forecaster_idxs[i]` being `NA` means there is no match.

File: R-notebooks/covidhub_evaluation.Rmd
Comment (JedGrabman) [reply R-notebooks/covidhub_evaluation.Rmd:194]: It should be, but it seemed out of scope for this PR and I'm not 100% on what the value should be (but I *think* I just need to check for 4 weeks back).

File: R-notebooks/covidhub_evaluation.Rmd
Comment (JedGrabman) [reply R-notebooks/covidhub_evaluation.Rmd:73]: `predictions_cards` is a list of lists. Each inner list (i.e. `predictions_cards[[i]]`) is specific to a single forecaster. The assumption is that an empty list will not be saved, and that all inner lists will have at least 1 element, with each element corresponding to the same forecaster. Therefore, checking the 1st element seems appropriate, because we cannot guarantee that there are any additional elements.

File: R-notebooks/covidhub_evaluation.Rmd
Comment (JedGrabman) [reply R-notebooks/covidhub_evaluation.Rmd:92]: Dates that we don't have data saved for. They are new in the sense that we will find any new forecasts that were made since the last time we ran the notebook. Technically, it includes old dates as well if they were non-comparable, because we didn't save that data. But, we filter those out when we find comparable dates.

I can add a comment.

File: R-notebooks/covidhub_evaluation.Rmd
Comment (JedGrabman) [reply R-notebooks/covidhub_evaluation.Rmd:97]: I will update

File: R-notebooks/covidhub_evaluation.Rmd
Comment (sgsmob) [reply R-notebooks/covidhub_evaluation.Rmd:92]: Please do

File: R-notebooks/covidhub_evaluation.Rmd
Comment (sgsmob) [inline R-notebooks/covidhub_evaluation.Rmd:107]: can be clearer as `as_date(setdiff(comparable_forecast_dates, seen_dates[[forecaster_idx]]))`

**Review by sgsmob — APPROVED** [review]
After addressing the few remaining comments, should be good to go.

File: R-notebooks/covidhub_evaluation.Rmd
Comment (sgsmob) [reply R-notebooks/covidhub_evaluation.Rmd:73]: Oops nevermind, I misread this. It is ok as is.

File: R-notebooks/covidhub_evaluation.Rmd
Comment (JedGrabman) [reply R-notebooks/covidhub_evaluation.Rmd]: Sorry, forgot to push before. Updated.

**Comment by JedGrabman** [comment]
@krivard I think this is good to go (pending your approval).

**Pull request merged**
